enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -5587,50 +5775,88 @@ class AdjointGenerator
+                isSum = true;
+              }
             }
           }
-        }
-        if (!isSum) {
-          llvm::errs() << *gutils->oldFunc << "\n";
-          llvm::errs() << *gutils->newFunc << "\n";
-          llvm::errs() << " call: " << call << "\n";
-          llvm::errs() << " unhandled mpi_allreduce op: " << *orig_op << "\n";
-          report_fatal_error("unhandled mpi_allreduce op");
+          if (!isSum) {
+            llvm::errs() << *gutils->oldFunc << "\n";
+            llvm::errs() << *gutils->newFunc << "\n";
+            llvm::errs() << " call: " << call << "\n";
+            llvm::errs() << " unhandled mpi_allreduce op: " << *orig_op << "\n";
+            report_fatal_error("unhandled mpi_allreduce op");
+          }
         }
 
-        Value *shadow_recvbuf =
-            lookup(gutils->invertPointerM(orig_recvbuf, Builder2), Builder2);
+        Value *shadow_recvbuf = gutils->invertPointerM(orig_recvbuf, Builder2);
+        if (!forwardMode)
+          shadow_recvbuf = lookup(shadow_recvbuf, Builder2);
         if (shadow_recvbuf->getType()->isIntegerTy())
           shadow_recvbuf = Builder2.CreateIntToPtr(
               shadow_recvbuf, Type::getInt8PtrTy(call.getContext()));
-        Value *shadow_sendbuf =
-            lookup(gutils->invertPointerM(orig_sendbuf, Builder2), Builder2);
+
+        Value *shadow_sendbuf = gutils->invertPointerM(orig_sendbuf, Builder2);
+        if (!forwardMode)
+          shadow_sendbuf = lookup(shadow_sendbuf, Builder2);
         if (shadow_sendbuf->getType()->isIntegerTy())
           shadow_sendbuf = Builder2.CreateIntToPtr(
               shadow_sendbuf, Type::getInt8PtrTy(call.getContext()));
 
-        Value *count = lookup(gutils->getNewFromOriginal(orig_count), Builder2);
-        Value *datatype =
-            lookup(gutils->getNewFromOriginal(orig_datatype), Builder2);
-        Value *root = lookup(gutils->getNewFromOriginal(orig_root), Builder2);
+        Value *count = gutils->getNewFromOriginal(orig_count);
+        if (!forwardMode)
+          count = lookup(count, Builder2);
+
+        Value *datatype = gutils->getNewFromOriginal(orig_datatype);
+        if (!forwardMode)
+          datatype = lookup(datatype, Builder2);
+
+        Value *op = lookup(gutils->getNewFromOriginal(orig_op), Builder2);
+        if (!forwardMode)
+          op = lookup(op, Builder2);
+
+        Value *root = gutils->getNewFromOriginal(orig_root);
+        if (!forwardMode)
+          root = lookup(root, Builder2);
+
         Value *comm = lookup(gutils->getNewFromOriginal(orig_comm), Builder2);
+        if (!forwardMode)
+          comm = lookup(comm, Builder2);
 
         Value *rank = MPI_COMM_RANK(comm, Builder2, root->getType());
 
+        if (forwardMode) {
+          Value *args[] = {
+              /*sendbuf*/ shadow_sendbuf,
+              /*recvbuf*/ shadow_recvbuf,
+              /*count*/ count,
+              /*datatype*/ datatype,
+              /*op*/ op,
```

Comment:
I'm not convinced the regular op here is ok. For sum sure , but what if this is produce/max

Reply:
we are now catching non-sum ops, just like in reverse mode